### DISCUSSION TITLE
Newbie, Some :dep work but others don't

### DISCUSSION THREAD

**Discussion by z1fire** [evcxr/evcxr#312] [Q&A] [answered]
So I'm try to figure this all out. I've got evcxr_jupyter installed and i can run some rust in it. Some crates will work and some won't. I'm following a tutorial at https://towardsdatascience.com/rust-the-next-big-thing-in-data-science-319a03305883 which has you importing the polars crate. Several other tutorials has polars working but I can't get it to work right to even play around with it.  

![Screenshot 2023-08-21 075034](https://github.com/evcxr/evcxr/assets/80300887/661a5bfb-60f3-4cc9-a031-5334a471c9db)

**Comment by z1fire** [comment]
![image](https://github.com/evcxr/evcxr/assets/80300887/357f824b-939c-4b3c-a24e-e59b12a41033)
![image](https://github.com/evcxr/evcxr/assets/80300887/e311ef21-1192-4636-a871-adf0b50905e6)
I attached some more images so far it seems to be just an issue with me importing the polars crate, ive updated my version of rust and still no luck. the other crates are imported just fine

**Comment by davidlattimore** [comment, marked as answer]
It looks like evcxr_jupyter isn't reporting the error that's happening as well as the evcxr repl is. That's something that should definitely be fixed. Doing the same `:dep` line in the repl reports:

```
A compilation error was found in code we generated.
Ideally this shouldn't happen. Type :last_error_json to see details.
error[E0034]: multiple applicable items in scope
  --> /home/david/.cargo/registry/src/index.crates.io-6f17d22bba15001f/polars-arrow-0.28.0/src/array/get.rs:31:17
   |
31 |         if self.is_null_unchecked(item) {
   |                 ^^^^^^^^^^^^^^^^^ multiple `is_null_unchecked` found
   |
note: candidate #1 is defined in an impl of the trait `IsValid` for the type `A`
  --> /home/david/.cargo/registry/src/index.crates.io-6f17d22bba15001f/polars-arrow-0.28.0/src/is_valid.rs:31:5
   |
31 |     unsafe fn is_null_unchecked(&self, i: usize) -> bool {
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   = note: candidate #2 is defined in an impl of the trait `arrow2::array::Array` for the type `arrow2::array::PrimitiveArray<T>`
help: disambiguate the method for candidate #1
   |
31 |         if IsValid::is_null_unchecked(&self, item) {
   |            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
help: disambiguate the method for candidate #2
   |
31 |         if arrow2::array::Array::is_null_unchecked(&self, item) {
   |            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

```

From that error, my guess would be that some time since polars 0.28.0 was published, a newer version of the `arrow2` crate added the `is_null_unchecked` method. You've probably got two options:

1) You could try a newer version of polars - although that may mean that it won't exactly match the article you're following.
2) You could experiment with older versions of `arrow2` to see if you can make things work.

I just had a quick try of option 2

```
:dep arrow2 = { version = "=0.17.0" }
:dep polars = {version = "0.28.0"}
```

That appeared to complete without error.

**Reply by z1fire** [reply]
Oh man!!!!! specifying the arrow2 version worked! I had previously tried using the most recent version of polars but I think you are right about arrow2 being updated. I was able to import everything I need now and even start the data analysis process.
![image](https://github.com/evcxr/evcxr/assets/80300887/b85df744-250e-4966-bc7d-3ec7269eb0a6)
Thank you soooo much!

Now I just have to ask will there be a point in the future based on what you now know about the issue with polars and arrow2 that we will be able to import the newest version of polars to use?

**Reply by davidlattimore** [reply]
Glad I could help :)

If I do `:dep polars`, which will get the latest version of polars, it works for me. Well, by works, I mean it doesn't give any errors. I haven't tried to do anything with it. Does it not build for you?

If you're not sure if it's an evcxr specific problem, you can also try building things outside of evcxr. e.g.

```sh
cargo new test-polars
cd test-polars
cargo add polars
cargo build
```

**Comment by z1fire** [comment]
Now I don't know what is going on, I started a new workbook to reproduce the issue I was having importing the polars crate. I used :dep polars which I had tried earlier and this time it imported just fine.  Is it caching the crate somehow where I imported the earlier version with the earlier version of arrow2?

**Reply by davidlattimore** [reply]
In a new notebook, you should have completely fresh state and it should be using the latest version of polars and the latest semver-compatible version of arrow2.

If you're unsure what it's doing you can run `:last_compile_dir`. In the terminal, you can then `cd` into that directory and look at the `Cargo.toml` to see what is being done. To see exact versions of crates being used, you can run `cargo tree`

**Comment by z1fire** [comment]
I'm now thinking that updating my version of rust was the answer all along.  I went through all the compiles through the day. earlier this morning it before I updated rust it showed a fail on polars 0.28.0. and now I've tried :dep polars for the newest version a couple times and its worked each time.  I can't reproduce the issues I had.  I can say that earlier today I tried everything as far as versions go. It must have been an issue with the version of rust I was running.  That's only a guess though. I'm still too new to make that call.

This is from this morning
![image](https://github.com/evcxr/evcxr/assets/80300887/360ddfd7-66da-418c-985d-56ccb77f82f2)

This is now.
![image](https://github.com/evcxr/evcxr/assets/80300887/b02f414a-99b3-498f-8c01-fb9dc3b8e0ec)
